app/uk/gov/hmrc/agentclientrelationships/services/RelationshipsService.scala

```diff
@@ -104,73 +127,115 @@ class RelationshipsService @Inject()(gg: GovernmentGatewayProxyConnector,
           updateGgSyncStatus(SyncStatus.IncompleteInputParams)
-        case ex                              =>
-          Logger.warn(s"Creating GG record failed", ex)
+        case NonFatal(ex) =>
+          Logger.warn(s"Creating GG record failed for ${arn.value}, ${mtdItId.value}", ex)
           updateGgSyncStatus(SyncStatus.Failed)
           if (failIfAllocateAgentInGGFails) Future.failed(new Exception("RELATIONSHIP_CREATE_FAILED_GG"))
           else Future.successful(())
       }
+  }
+
+  def createRelationship(arn: Arn,
+                         mtdItId: MtdItId,
+                         eventualAgentCode: Future[AgentCode],
+                         oldReferences: Set[SaAgentReference],
+                         failIfCreateRecordFails: Boolean,
+                         failIfAllocateAgentInGGFails: Boolean
+                        )
+                        (implicit ec: ExecutionContext, hc: HeaderCarrier, auditData: AuditData): Future[Unit] = {
+
+    auditData.set("AgentDBRecord", false)
+    auditData.set("enrolmentDelegated", false)
+    auditData.set("etmpRelationshipCreated", false)
+
+    def createRelationshipRecord: Future[Unit] = {
+      val record = RelationshipCopyRecord(arn.value, mtdItId.value, MtdItIdType, Some(oldReferences))
+      relationshipCopyRepository.create(record)
+        .map(_ => auditData.set("AgentDBRecord", true))
+        .recoverWith {
+          case NonFatal(ex) =>
+            Logger.warn(s"Inserting relationship record into mongo failed for ${arn.value}, ${mtdItId.value}", ex)
+            if (failIfCreateRecordFails) Future.failed(new Exception("RELATIONSHIP_CREATE_FAILED_DB"))
+            else Future.successful(())
+        }
+    }
 
     for {
       _ <- createRelationshipRecord
-      _ <- createEtmpRecord(mtdItId)
-      _ <- createGgRecord(mtdItId)
+      _ <- createEtmpRecord(arn, mtdItId)
+      _ <- createGgRecord(arn, mtdItId, eventualAgentCode, failIfAllocateAgentInGGFails)
     } yield ()
   }
 
-  private def intersection[A](cesaIds: Seq[A])(mappingServiceCall: => Future[Seq[A]])(implicit hc: HeaderCarrier): Future[Set[A]] = {
+  private def recoverRelationshipCreation(
+    relationshipCopyRecord: RelationshipCopyRecord,
+    arn: Arn, mtdItId: MtdItId,
+    eventualAgentCode: Future[AgentCode])(implicit ec: ExecutionContext, hc: HeaderCarrier, auditData: AuditData): Future[Unit] = {
+
+    lockService.tryLock(arn, mtdItId) {
+      def recoverEtmpRecord() = createEtmpRecord(arn, mtdItId)
+
+      def recoverGgRecord() = createGgRecord(arn, mtdItId, eventualAgentCode, failIfAllocateAgentInGGFails = false)
+
+      (relationshipCopyRecord.needToCreateEtmpRecord, relationshipCopyRecord.needToCreateGgRecord) match {
+        case (true, true) =>
+          for {
+            _ <- recoverEtmpRecord()
+            _ <- recoverGgRecord()
+          } yield ()
+        case (false, true) =>
+          recoverGgRecord()
```

Comment:
We don't log a warning here because this is an expected case. The other 2 we don't expect to happen.
Could log an info if you like?